subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -106,7 +102,12 @@ public JavaCompile() {
         ObjectFactory objectFactory = getObjectFactory();
         compileOptions = objectFactory.newInstance(CompileOptions.class);
         modularity = objectFactory.newInstance(DefaultModularitySpec.class);
-        javaCompiler = objectFactory.property(JavaCompiler.class);
+        JavaToolchainService javaToolchainService = getJavaToolchainService();
```

Comment:
Seeing this multiple times now makes me think all of these provider setups (`JavaExec`, `JavaCompile`, `Javadoc`) should live in the same static utils class.

Reply:
As noted above, we'll fix this in the follow-up PRs.